lib/helpers.bash

```diff
@@ -43,20 +45,25 @@ function _load_global_bash_it_files() {
   fi
 }
 
-# Function for reloading aliases
-function reload_aliases() {
-  _load_bash_it_files "aliases"
+function _make_reload_alias() {
+  printf %s '\
+  _bash_it_list_bash_it_files_return=() ;\
+  _list_bash_it_files '"$1"' ;\
+  for _bash_it_config_file in "${_bash_it_list_bash_it_files_return[@]}"; do \
+    . "$_bash_it_config_file" ;\
+  done ;\
+  unset _bash_it_list_bash_it_files_return ;\
+  unset _bash_it_config_file'
```

Comment:
For reference, here's what my `_make_reload_alias` looks like (with debugging code):

```bash
function _make_reload_alias() {
  printf %s '
  echo '"$1"';
  for _bash_it_enabled_config_file in $(_list_bash_it_files "'$1'"); do
    echo "reloading ${_bash_it_enabled_config_file}";
    . "$_bash_it_enabled_config_file";
  done;
  echo "done";
  unset _bash_it_enabled_config_file'
}
```

The alias ends up looking something like this:

```bash
alias reload_aliases='
  echo aliases;
  echo $(_list_bash_it_files aliases);
  for _bash_it_enabled_config_file in $(_list_bash_it_files "aliases"); do
    echo "reloading ${_bash_it_enabled_config_file}";
    . "$_bash_it_enabled_config_file";
  done;
  echo "done";
  unset _bash_it_enabled_config_file'
```

Reply:
Of course, it took posting this to realize a possible solution. This works:

```bash
function _make_reload_alias() {
  printf %s '
  if [ -d "${BASH_IT}/'$1'/enabled" ]; then
    for _bash_it_enabled_config_file in ${BASH_IT}/'$1'/enabled/*.bash; do
      if [ -f "$_bash_it_enabled_config_file" ]; then
        . "$_bash_it_enabled_config_file";
      fi
    done;
    unset _bash_it_enabled_config_file
  fi'
}
```

I'm sure it could be made better though.

Reply:
Is the `$1` in your code in `${BASH_IT}/'$1'/enabled/*.bash` used for selecting the `.bash_it/aliases/enabled` directory? If that is the case, then this is not going to work anymore with newer versions of Bash-it. I had moved all enabled components to a central `${BASH_IT}/enabled` directory, to allow loading in a well-defined order.

Existing installations should automatically be migrated to the new format when enabling a new component - after that, directories like `${BASH_IT}/aliases/enabled` should be empty...

You would have to do the same for the global `${BASH_IT}/enabled` directory, but this time with a `*.alias.bash` glob.

Reply:
@nwinkler Good to know, I was wondering what the use of _load_global_bash_it_files was.

I'll patch it as well.

Reply:
@nwinkler yup – my plan was to get this working, then add in a check for `${BASH_IT}/enabled/*.${1}.bash` as well so we can add the reload of specific types back.

That said, I also thought about dropping these `reload_` functions/aliases all together so we can just clean this up.

Reply:
> Of course, it took posting this to realize a possible solution. This works:
> 
> ```shell
> function _make_reload_alias() {
>   printf %s '
>   if [ -d "${BASH_IT}/'$1'/enabled" ]; then
>     for _bash_it_enabled_config_file in ${BASH_IT}/'$1'/enabled/*.bash; do
>       if [ -f "$_bash_it_enabled_config_file" ]; then
>         . "$_bash_it_enabled_config_file";
>       fi
>     done;
>     unset _bash_it_enabled_config_file
>   fi'
> }
> ```
> 
> I'm sure it could be made better though.

In my experience, multiline aliases are quite finicky. At least my bash version (4.4.23) doesn't seem to handle them very well - it'll print mysterious syntax errors, or worse, fail to execute some lines silently. I had to resort to escaping newlines to make things work properly.

Reply:
Here's the latest cleaned version of this function:

```
function _make_reload_alias() {
  local global_family="$1"
  local subdirectory="$2"

  printf %s '
  for _bash_it_config_file in $(_list_global_bash_it_files '"$global_family"'); do
    . "${BASH_IT}/$_bash_it_config_file"
  done ;\
  for _bash_it_config_file in $(_list_bash_it_files '"$subdirectory"'); do
    . "${BASH_IT}/$_bash_it_config_file"
  done ;\
  unset _bash_it_config_file'
}
```
Without the first backslash `;\`, Bash only executes the first for loop and then stops silently.

Without the second `;\`, `unset` is not run, again silently.

I'm not sure why that is - multiline aliases seem to work fine in general, but they might not be super-well tested since in most cases using a function is a superior alternative. Could this be a Bash bug?

Reply:
After a bit more digging, I'm inclined to think that this is a Bash bug.

I've boiled it down to this:

```
alias bug='echo one
a="$(pwd)"
echo two'

bug
```

This prints 'one' on bash 4.4 when e.g. reload_aliases is run.

It seems like Bash stops executing a multiline alias right after exiting the statement or block in which command substitution happened.

Command substitution seems to be the only culprit here - other kinds of substitutions and subshells don't trigger the issue.

I have more on this, but I've got to go now. Will carry on the analysis later.

Reply:
I was mulling this over on the way to work this morning and I'm wondering if there isn't something with aliases and subshell execution – which actually gives me an idea. What about abstracting the functionality into its own scripts file that is source'd to load things. Something like:

```bash
alias reload_plugins='source ${BASH_IT}/scripts/reload.bash "plugins"'
```

We could wrap it inside `_make_reload_alias` to reduce the duplication.

Reply:
> > I had moved all enabled components to a central ${BASH_IT}/enabled directory, to allow loading in a well-defined order.
> 
> I don't think that's enough to guarantee a well-defined load order. AFAIK, globing order is locale-dependent, and it's really easy for even English-speaking users to get bitten by this: for instance, '.' does not get collated the same way between the C and en_US.UTF-8 locales.

The well-defined order is described in the https://github.com/Bash-it/bash-it/blob/master/DEVELOPMENT.md document. It's primarily based on numbers/priorities - the different character sets shouldn't come into play with this.

Default priorities are:

* Aliases: 150
* Plugins: 250
* Completions: 350

The enabled files are symlinked into the global `enabled` folder with the priority as a prefix, e.g. `150---git.aliases.bash`.

This ensures that all aliases are loaded first, then plugins, then completions, but you can make exceptions if needed. Examples are:

* The `nvm` plugin needs to be loaded before the `node` plugin, hence `nvm` has a load priority of `225` instead of the standard `250`.
* The `alias-completion` plugin needs to be loaded after all completions have been loaded, hence `alias-completion` has a load priority of `365` instead of the standard `250`.

When changing the reload code, this logic needs to stay in place, otherwise the results will be undefined...

Reply:
> I was mulling this over on the way to work this morning and I'm wondering if there isn't something with aliases and subshell execution – which actually gives me an idea. What about abstracting the functionality into its own scripts file that is source'd to load things. Something like:
> 
> ```shell
> alias reload_plugins='source ${BASH_IT}/scripts/reload.bash "plugins"'
> ```
> 
> We could wrap it inside `_make_reload_alias` to reduce the duplication.

I like this idea! I assume that sourcing the files from a script does not have the same `declare` issue that we have with sourcing from a function - which this PR is all about...

Reply:
I like this idea as well!

I ran some tests yesterday on various versions of Bash, and the results were not pretty:

- The command substitution bug impacts all 4.x versions, but neither 3.2 nor 5.0-rc
- The hackish fix that consists in adding `;\` works on all 4.x version I tested, except for 4.0.

So it looks like something got broken in 4.0, and partially fixed later until 5.0 fixed it completely.

Sidestepping the issue is definitely the sane thing to do.